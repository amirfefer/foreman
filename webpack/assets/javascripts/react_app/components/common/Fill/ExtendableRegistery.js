const components = {};

export const add = (SlotId, fillId, component, weight, overrideProps) => {
  if (components[SlotId] === undefined) {
    components[SlotId] = [];
  }

  component = component || overrideProps;

  components[SlotId].push({ fillId, component, weight });
};

export const remove = (SlotId, fillId) => {
  const slotItems = components[SlotId];

  components[SlotId] = slotItems.filter(item => item.fillId !== fillId);
};

export const getComponentsBySlotId = id => components[id];
