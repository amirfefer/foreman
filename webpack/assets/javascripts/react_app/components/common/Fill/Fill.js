import React from 'react';
import PropTypes from 'prop-types';

class Fill extends React.Component {
  componentDidMount() {
    const {
      children,
      overrideProps,
      registerFillComponent,
      id,
      weight,
      fillId,
    } = this.props;

    registerFillComponent(id, overrideProps, fillId, children, weight);
  }
  componentWillUnmount() {
    const { id, unregisterFillComponent, fillId } = this.props;
    unregisterFillComponent(id, fillId);
  }
  render() {
    return null;
  }
}

Fill.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.object]),
  registerFillComponent: PropTypes.func.isRequired,
  unregisterFillComponent: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  weight: PropTypes.number.isRequired,
  fillId: PropTypes.string.isRequired,
  overrideProps: PropTypes.object,
};

Fill.defaultProps = {
  children: undefined,
  overrideProps: undefined,
};

export default Fill;
