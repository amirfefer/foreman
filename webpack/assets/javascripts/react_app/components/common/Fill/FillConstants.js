export const REGISTER_FILL = 'REGISTER_FILL';

export const REMOVE_FILLED_COMPONENT = 'REMOVE_FILLED_COMPONENT';
