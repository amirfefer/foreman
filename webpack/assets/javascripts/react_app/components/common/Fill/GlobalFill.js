import { registerFillComponent } from '../Fill/FillActions';
import store from '../../../redux';

export const AddGlobalFill = (slotId, fillId, component, weight) => {
  store.dispatch(
    registerFillComponent(slotId, undefined, fillId, component, weight)
  );
};
