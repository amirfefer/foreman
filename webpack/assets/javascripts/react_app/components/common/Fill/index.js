import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './FillActions';
import reducer from './FillReducer';

import Fill from './Fill';

// map state to props
const mapStateToProps = state => ({
  components: state.extendable,
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { extendable: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Fill);
