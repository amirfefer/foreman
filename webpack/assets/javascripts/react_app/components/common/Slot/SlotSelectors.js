import { createSelector } from 'reselect';
import { orderBy } from 'lodash';
import { getComponentsBySlotId } from '../Fill/ExtendableRegistery';

export const selectAllFills = slotId => getComponentsBySlotId(slotId);

export const selectComponentByWeight = createSelector(
  selectAllFills,
  fills => orderBy(fills, ['weight'], ['desc']).map(c => c.component)
);

export const selectMaxComponent = createSelector(
  selectComponentByWeight,
  fills => fills[0]
);

export const selectRegisteredFills = (state, slotId) =>
  state.extendable.RegisteredComponent[slotId] || {};
