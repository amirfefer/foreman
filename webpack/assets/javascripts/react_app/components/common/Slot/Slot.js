import { cloneElement, isValidElement } from 'react';
import uuidV1 from 'uuid/v1';
import PropTypes from 'prop-types';
import { selectComponentByWeight, selectMaxComponent } from './SlotSelectors';

const Slot = ({ fills, id, multi, children, ...props }) => {
  const addProps = object => {
    if (isValidElement(object)) {
      return cloneElement(object, { ...props, key: uuidV1() });
    }

    if (children === undefined) {
      throw new Error('Slot with override props must having children');
    }
    if (typeof object !== 'object') {
      throw new Error(
        'Slot with override props must having props object instead of JSX'
      );
    }
    return cloneElement(children, { ...props, ...object, key: uuidV1() });
  };

  if (Object.keys(fills).length) {
    if (multi) {
      return selectComponentByWeight(id).map(component => addProps(component));
    }
    return addProps(selectMaxComponent(id));
  }
  if (children) return children;
  return null;
};

Slot.propTypes = {
  fills: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired,
  multi: PropTypes.bool,
  children: PropTypes.node,
};

Slot.defaultProps = {
  multi: false,
  children: undefined,
};

export default Slot;
