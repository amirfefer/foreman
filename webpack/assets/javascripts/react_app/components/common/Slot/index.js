import { connect } from 'react-redux';
import { selectRegisteredFills } from './SlotSelectors';
import Slot from './Slot';

// map state to props
const mapStateToProps = (state, ownProps) => ({
  fills: selectRegisteredFills(state, ownProps.id),
});

// export connected component
export default connect(mapStateToProps)(Slot);
