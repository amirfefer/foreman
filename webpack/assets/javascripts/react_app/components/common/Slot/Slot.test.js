import toJson from 'enzyme-to-json';
import React from 'react';

import IntegrationTestHelper from '../../../common/IntegrationTestHelper';
import Fill from '../Fill';
import Slot from './';
import FillReducer from '../Fill/FillReducer';

const combinedReducers = { extendable: FillReducer };
const SlotComponent = ({ text }) => <div>{text}</div>;

describe('Slot-Fill', () => {
  const integrationTestHelper = new IntegrationTestHelper(combinedReducers);

  it('multiple fills', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill id="slot-1" fillId="some-key-1" weight={100}>
          <div> Extendable component 1 </div>
        </Fill>
        <Fill id="slot-1" fillId="some-key-2" weight={200}>
          <span> Extendable component 2 </span>
        </Fill>

        <Slot id="slot-1" multi>
          <div> defaul value </div>
        </Slot>
      </React.Fragment>
    );

    wrapper.update();
    expect(toJson(wrapper.find('Slot'))).toMatchSnapshot();
  });
  it('no nultiple fills', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill id="slot-2" fillId="some-key-1" weight={100}>
          <div> Extendable component 1 </div>
        </Fill>
        <Fill id="slot-2" fillId="some-key-1" weight={200}>
          <span> Extendable component 2 </span>
        </Fill>

        <Slot id="slot-2" multi={false}>
          <div> defaul value </div>
        </Slot>
      </React.Fragment>
    );

    wrapper.update();
    expect(toJson(wrapper.find('Slot'))).toMatchSnapshot();
  });
  it('props fill', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill
          overrideProps={{ text: 'This is given by a prop' }}
          id="slot-3"
          fillId="some-key-1"
          weight={100}
        />

        <Slot id="slot-3" multi={false}>
          <SlotComponent />
        </Slot>
      </React.Fragment>
    );

    wrapper.update();
    expect(toJson(wrapper.find('Slot'))).toMatchSnapshot();
  });
  it('default slot', () => {
    const wrapper = integrationTestHelper.mount(
      <Slot id="slot-4" multi>
        <SlotComponent text="Default Value" />
      </Slot>
    );

    wrapper.update();
    expect(toJson(wrapper.find('Slot'))).toMatchSnapshot();
  });
  it('multiple slot should be forbidden when override props', () => {
    const wrapper = expect(
      integrationTestHelper
        .mount(
          <React.Fragment>
            <Fill
              overridProps={{ text: 'This is given by a prop' }}
              id="slot-5"
              fillId="some-key-1"
              weight={100}
            />
            <Fill
              overridProps={{ text: 'This is given by a prop' }}
              id="slot-5"
              fillId="some-key-2"
              weight={200}
            />
            <Slot id="slot-5" multi>
              <SlotComponent text="Default Value" />
            </Slot>
          </React.Fragment>
        )
        .toThrowError(new Error())
    );
  });
});
