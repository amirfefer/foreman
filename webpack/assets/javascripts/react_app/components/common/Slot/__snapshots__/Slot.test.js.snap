// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot-Fill default slot 1`] = `
<Slot
  dispatch={[Function]}
  id="slot-4"
  multi={true}
  shouldUpdate={false}
>
  <SlotComponent
    text="Default Value"
  >
    <div>
      Default Value
    </div>
  </SlotComponent>
</Slot>
`;

exports[`Slot-Fill multiple fills 1`] = `
<Slot
  dispatch={[Function]}
  id="slot-1"
  multi={true}
  shouldUpdate={true}
>
  <div
    0={
      <span>
         Extendable component 2 
      </span>
    }
    1={
      <div>
         Extendable component 1 
      </div>
    }
  >
     defaul value 
  </div>
</Slot>
`;

exports[`Slot-Fill no nultiple fills 1`] = `
<Slot
  dispatch={[Function]}
  id="slot-2"
  multi={false}
  shouldUpdate={true}
>
  <span
    dispatch={[Function]}
  >
     Extendable component 2 
  </span>
</Slot>
`;

exports[`Slot-Fill props fill 1`] = `
<Slot
  dispatch={[Function]}
  id="slot-3"
  multi={false}
  shouldUpdate={true}
>
  <SlotComponent
    text="This is given by a prop"
  >
    <div>
      This is given by a prop
    </div>
  </SlotComponent>
</Slot>
`;
